manager/src/main/java/org/openremote/manager/mqtt/KeycloakAuthorizatorPolicy.java

```diff
@@ -0,0 +1,111 @@
+        MqttConnector.MqttConnection connection = mqttConnector.getConnection(clientId);
+        if (connection == null) {
+            LOG.info("No connection found for clientId: " + clientId);
+            return false;
+        }
+
+        if (!connection.username.equals(username)) {
+            LOG.info("Username mismatch");
+            return false;
+        }
+
+        if (topic.isEmpty() || topic.getTokens().size() != 3) {
+            LOG.info("Topic may not be empty and should have the following format: assets/{assetId}/{attributeName}");
+            return false;
+        }
+
+        if (!topic.headToken().toString().equals(ASSETS_TOPIC)) {
+            LOG.info("Topic should have the following format: assets/{assetId}/{attributeName}");
+            return false;
+        }
+
+        String assetId = topic.getTokens().get(1).toString();
+        Asset asset = assetStorageService.find(assetId);
+        if (asset == null) {
+            LOG.info("Asset not found for assetId: " + assetId);
+            return false;
+        }
+
+        if (!asset.getRealm().equals(connection.realm)) {
+            LOG.info("Asset not in same clientId");
+            return false;
+        }
+
+        String attributeName = topic.getTokens().get(2).toString();
+        if(!asset.getAttribute(attributeName).isPresent()) {
+            LOG.info("Attribute not found: " + attributeName);
+            return false;
+        }
+
+        try {
+            AccessToken accessToken = AdapterTokenVerifier.verifyToken(connection.accessToken, identityProvider.getKeycloakDeployment(connection.realm, KEYCLOAK_CLIENT_ID));
+            return identityProvider.canSubscribeWith(new AccessTokenAuthContext(connection.realm, accessToken), new TenantFilter(connection.realm), roles);
+        } catch (VerificationException e) {
+            if (e instanceof TokenNotActiveException) {
+                String suppliedClientSecret = new String(connection.password, StandardCharsets.UTF_8);
+                connection.accessToken = identityProvider.getKeycloak().getAccessToken(connection.realm, new ClientCredentialsAuthFrom(connection.username, suppliedClientSecret)).getToken();
+                try {
+                    AccessToken accessToken = AdapterTokenVerifier.verifyToken(connection.accessToken, identityProvider.getKeycloakDeployment(connection.realm, KEYCLOAK_CLIENT_ID));
+                    return identityProvider.canSubscribeWith(new AccessTokenAuthContext(connection.realm, accessToken), new TenantFilter(connection.realm), roles);
+                } catch (VerificationException verificationException) {
+                    LOG.log(Level.INFO, "Couldn't verify token", verificationException);
+                    return false;
+                }
+            } else {
+                LOG.log(Level.INFO, "Couldn't verify token", e);
+                return false;
+            }
+        }
+    }
+}
```

Comment:
The idea we discussed was to use the existing infrastructure, so this code should just create `EventSubscription`s and push through a route in `ClientEventService` that is used by Websocket and MQTT clients.